### DISCUSSION TITLE
How to connect to self-hosted GitLab?

### DISCUSSION THREAD

**Discussion by berislavlopac** [windmill-labs/windmill#1835] [Questions]
I'm trying to connect to a self-hosted instance of GitLab, but when adding a new resource it always try to get authorisation from gitlab.com. The GitLab resource type has the `baseUrl` field, but it cannot be edited when creating a new resource.

Any help appreciated! 😊

**Comment by rubenfiszel** [comment]
The baseUrl is editable, are you sure you cannot edit that field (I checked and it was working fine)

**Reply by berislavlopac** [reply]
That's what I was expecting. But when I go to Resources -> Add a Resource/API -> GitLab (in the "OAuth APIs" section"), it only provides a number of "Scopes" fields; clicking on "Connect" takes me directly to gitlab.com, which asks for autorisation. 🤷‍♂️ 

How exactly do I edit the `baseUrl` field? 🤔

**Reply by rubenfiszel** [reply]
Oh yeah that's completely different since that's Oauth. You can configure custom OAuth on a dedicated instance/plan or using self-hosted but we can only install our own OAuth for our own apps in the multi-tenant instance.

**Reply by berislavlopac** [reply]
I don't understand - different from what? Is there a way to connect to a self-hosted GitLab at all? Is it maybe a paid feature?

**Reply by rubenfiszel** [reply]
Windmill has native handling of OAuth for some resources. This enable to get the access token and refresh the token using the OAuth scheme in the background. OAuth require 2 parts, a client and an app setup in the target, here gitlab. The client at the moment need to be configured instance-wide, and the app is, for the cloud, an official windmill app configured following: those instructions: https://docs.gitlab.com/ee/integration/oauth_provider.html. 
The client can be configured on dedicated hosted plans and self-hosted installation to point to your own configured app.

However, some resource types make sense as both oauth managed resources AND manually written resources. This is the case for gitlab so I just did a commit so that it will be displayed in both OAuth resources and non OAuth resources. That should enable you to put an API key in it manually.

**Reply by berislavlopac** [reply]
Thank you Ruben, this works like a charm! 👍
